FeedStoreChallenge/CoreData/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,49 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				if let data = try ManagedCache.find(in: context) {
+					completion(.found(feed: data.local, timestamp: data.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				try ManagedCache.createNewOrUpdateStoredFeed(with: feed, at: timestamp, in: context)
+				if context.hasChanges {
+					try context.save()
+				}
```

Comment:
At this point `context` will definitely have changes, there's no need to check `context.hasChanges`.

Please remove `context.hasChanges`

```suggestion
				try context.save()
```